tpls/desul/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,491 @@
+  unsigned int tmp;
+  tmp = atomicMax(ptr, val);
+  tmp = std::max(tmp, val);
+  return tmp;
+}
+
+#pragma acc routine seq
+inline unsigned long long device_atomic_max_fetch(
+unsigned long long* ptr, const unsigned long long val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  unsigned long long tmp;
+  tmp = atomicMax(ptr, val);
+  tmp = std::max(tmp, val);
+  return tmp;
+}
+#endif
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_min_fetch(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  if(acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_min_fetch(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  *ptr = std::min(*ptr, val); tmp = *ptr;
+  return tmp;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
+#pragma acc routine seq
+inline int device_atomic_min_fetch(
+int* ptr, const int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  int tmp;
+  tmp = atomicMin(ptr, val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+
+#pragma acc routine seq
+inline unsigned int device_atomic_min_fetch(
+unsigned int* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  unsigned int tmp;
+  tmp = atomicMin(ptr, val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+
+#pragma acc routine seq
+inline unsigned long long device_atomic_min_fetch(
+unsigned long long* ptr, const unsigned long long val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  unsigned long long tmp;
+  tmp = atomicMin(ptr, val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+#endif
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_and_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { *ptr &= val; tmp = *ptr; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_or_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { *ptr |= val; tmp = *ptr; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_xor_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { *ptr ^= val; tmp = *ptr; }
+  return tmp;
+}
+//</editor-fold>
+
+//<editor-fold desc="device_atomic_{store,load}">
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,void> device_atomic_store(
+    T* const ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+#pragma acc atomic write
+    *ptr = val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder>
+acc_enable_if_supported_arithmetic_type<T,void> device_atomic_store(
+    T* const ptr, const T val, MemoryOrder, MemoryScopeDevice) {
+  device_atomic_store(ptr, val, MemoryOrderRelaxed(), MemoryScopeDevice()); 
```

Comment:
Does that mean the atomic write is actually sequentially consistent?

Reply:
No, OpenACC has a relaxed memory consistency model; the reason for having this implementation is that the current OpenACC atomics implementation uses the same `device_atomic_store()` API for both host and device.
Host may use memory consistency other than `MemoryOrderRelaxed`, but it is not possible for OpenACC to implement memory consistency models for atomic operations other than the relaxed consistency.
Therefore, the same relaxed consistency implementation is re-used for other Host-side `atomic_store()`.
(This is OK since the current OpenACC atomic implementation supports the host only in a serial execution mode.)